ts-client/src/amm/index.ts

```diff
@@ -0,0 +1,935 @@
+import { ASSOCIATED_PROGRAM_ID } from '@project-serum/anchor/dist/cjs/utils/token';
+import { calculateWithdrawableAmount, VaultState } from '@mercurial-finance/vault-sdk';
+import invariant from 'invariant';
+import { AccountsInfo, AmmImplementation, DepositQuote, PoolInformation, PoolState, WithdrawQuote } from './types';
+import { Amm, IDL as AmmIdl } from './idl';
+import { Vault, IDL as VaultIdl } from './vault-idl';
+import {
+  DEFAULT_SLIPPAGE,
+  DEVNET_COIN,
+  ERROR,
+  CURVE_TYPE_ACCOUNTS,
+  PROGRAM_ID,
+  SEEDS,
+  VAULT_PROGRAM_ID,
+  WRAPPED_SOL_MINT,
+} from './constants';
+import { StableSwap, SwapCurve, TradeDirection } from './curve';
+import { ConstantProductSwap } from './curve/constant-product';
+import {
+  calculateMaxSwapOutAmount,
+  calculateSwapQuote,
+  computeActualDepositAmount,
+  calculatePoolInfo,
+  getMaxAmountWithSlippage,
+  getMinAmountWithSlippage,
+  getOrCreateATAInstruction,
+  unwrapSOLInstruction,
+  wrapSOLInstruction,
+  getDepegAccounts,
+} from './utils';
+
+type AmmProgram = Program<Amm>;
+type VaultProgram = Program<Vault>;
+
+type Opt = {
+  cluster: Cluster;
+};
+
+const getPoolState = async (poolMint: PublicKey, program: AmmProgram) => {
+  const poolState = (await program.account.pool.fetchNullable(poolMint)) as PoolState;
+  invariant(poolState, `Pool ${poolMint.toBase58()} not found`);
+
+  const account = await program.provider.connection.getTokenSupply(poolState.lpMint);
+  invariant(account.value.amount, ERROR.INVALID_ACCOUNT);
+
+  return { ...poolState, lpSupply: new BN(+account.value.amount) };
+};
+
+const getVaultState = async (vaultMint: PublicKey, program: VaultProgram) => {
+  const vaultState = (await program.account.vault.fetchNullable(vaultMint)) as VaultState;
+  invariant(vaultState, `Vault ${vaultMint.toBase58()} not found`);
+
+  const account = await program.provider.connection.getTokenSupply(vaultState.lpMint);
+  invariant(account.value.amount, ERROR.INVALID_ACCOUNT);
+
+  return vaultState;
+};
+
+const getRemainingAccounts = (poolState: PoolState) => {
+  let accounts: Array<{
+    pubkey: PublicKey;
+    isWritable: boolean;
+    isSigner: boolean;
+  }> = [];
+  if ('stable' in poolState.curveType) {
+    if ('marinade' in poolState.curveType['stable'].depeg.depegType) {
+      accounts.push({
+        pubkey: CURVE_TYPE_ACCOUNTS.marinade,
+        isWritable: false,
+        isSigner: false,
+      });
+    }
+
+    if ('lido' in poolState.curveType['stable'].depeg.depegType) {
+      accounts.push({
+        pubkey: CURVE_TYPE_ACCOUNTS.solido,
+        isWritable: false,
+        isSigner: false,
+      });
+    }
+  }
+
+  return accounts;
+};
+
+const getAccountsBuffer = async ({
+  connection,
+  vaultA,
+  vaultB,
+  apyPda,
+  poolState,
+}: {
+  connection: Connection;
+  vaultA: VaultState;
+  vaultB: VaultState;
+  apyPda: PublicKey;
+  poolState: PoolState;
+}) => {
+  return await connection.getMultipleAccountsInfo([
```

Comment:
Redundant `await`. And, can you put a comment to indicate that the order of returned account info is the same order as `deserializeAccountsBuffer`

Reply:
ahh right